### PR TITLE
Custom authentication fix

### LINKED ISSUE #177: Custom Authentication is Broken
When we try to use custom authentication with `UsernamePasswordCredentials` class, we can only configure it by using XML configuration as below:

```
<security> 
    <credentials>com.hazelcast.security.UsernamePasswordCredentials</credentials> 
    <login-credentials> 
        <username>User</username> 
        <password>usrPassword</password> 
    </login-credentials> 
</security>
```
It doesn't allow to configure programmatically since the `UsernamePasswordCredentials` is internal. It is a public class in Java client.

### PR DESCRIPTION
**Pull request by asimarslan**
fixes #177

### CODE REVIEW

File: Hazelcast.Net/Hazelcast.Security/DefaultCredentialsFactory.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        
+        private ICredentials InitCredentials(ClientSecurityConfig securityConfig, GroupConfig groupConfig)
+        {
+            var credentials = securityConfig.GetCredentials();
+            if (credentials == null) {
+                var credentialsClassname = securityConfig.GetCredentialsClassName();
+                if (credentialsClassname != null) 
+                {
+                    try {
+//                        credentials = ClassLoaderUtil.newInstance(classLoader, credentialsClassname);
```
Comment (sancar) [inline Hazelcast.Net/Hazelcast.Security/DefaultCredentialsFactory.cs]: Why is this part commented out ? 
And I could not see a test with only `setCredentialsName` is used. Is there a test for this branch ?

File: Hazelcast.Net/Hazelcast.Client.Connection/ClientConnectionManager.cs [hunk 2]
```diff
@@ -239,11 +236,12 @@ public Task<ClientConnection> GetOrConnectAsync(Address address)
             {
                 if (!_pendingConnections.ContainsKey(address))
                 {
-                    var task = _client.GetClientExecutionService().Submit(() => GetOrConnect(address));
+                    var task = _client.GetClientExecutionService().Submit(() => GetOrConnect(address, isOwner:false));
                     task.ContinueWith(t =>
                     {
                         if (t.IsFaulted)
                         {
+                            //todo  AUTHENTICATE
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Connection/ClientConnectionManager.cs]: leftover, I'll clean it

File: Hazelcast.Net/Hazelcast.Client.Connection/ClientConnectionManager.cs [hunk 3]
```diff
@@ -269,54 +267,84 @@ private void CheckLive()
...
+                            throw new HazelcastException(string.Format("Node with address '{0}' was not found in the member list",
+                                result.address));
+                        }
+                        connection.Member = member;
+                    }
+                    break;
+                case AuthenticationStatus.CredentialsFailed:
+                    throw new AuthenticationException("Invalid credentials! Principal: " + ClientPrincipal);
+                case AuthenticationStatus.SerializationVersionMismatch:
+                    throw new InvalidOperationException("Server serialization version does not match to client");
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Connection/ClientConnectionManager.cs]: This exception is thrown when client cannot connect to server.

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientClusterService.cs [hunk 4]
```diff
@@ -317,19 +307,18 @@ private bool Connect(ICollection<IPEndPoint> triedAddresses)
...
                     FireConnectionEvent(LifecycleEvent.LifecycleState.ClientConnected);
-                    OwnerConnectionAddress = connection.GetAddress();
+                    OwnerConnectionAddress = ownerConnection.GetAddress();
                     return true;
                 }
                 catch (Exception e)
                 {
+                    if (exceptions.Count == MaxExceptionCount)
+                    {
+                        exceptions.RemoveAt(0);
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientClusterService.cs]: During connection retries, client may get multiple exceptions and not always the last exception is best. But we don't want an OOME during infinite retries and just keep last 10 of them. 
There is a good fit AggregateException to return multiple exceptions for such cases.

File: Hazelcast.Net/Hazelcast.Client/HazelcastClient.cs [hunk 5]
```diff
@@ -463,6 +470,7 @@ private void Start()
                 _loadBalancer.Init(GetCluster(), _config);
                 _partitionService.Start();
                 _statistics.Start();
+                _credentialsFactory.Destroy();
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client/HazelcastClient.cs]: 😱

**Pull request merged**